strings/base_collections_base.h

```diff
@@ -70,43 +140,66 @@ WINRT_EXPORT namespace winrt
 
             T Current() const
             {
-                if (m_current == m_end)
+                return m_owner->perform_shared([&]
                 {
-                    throw hresult_out_of_bounds();
-                }
+                    this->check_version(*m_owner);
 
-                if constexpr (!impl::is_key_value_pair<T>::value)
-                {
-                    return m_owner->unwrap_value(*m_current);
-                }
-                else
-                {
-                    return make<impl::key_value_pair<T>>(m_owner->unwrap_value(m_current->first), m_owner->unwrap_value(m_current->second));
-                }
+                    if (m_current == m_end)
+                    {
+                        throw hresult_out_of_bounds();
+                    }
+
+                    return current_value();
+                });
             }
 
-            bool HasCurrent() const noexcept
+            bool HasCurrent() const
             {
-                return m_current != m_end;
+                return m_owner->perform_shared([&]
+                {
+                    this->check_version(*m_owner);
+                    return m_current != m_end;
+                });
             }
 
-            bool MoveNext() noexcept
+            bool MoveNext()
             {
-                if (m_current != m_end)
+                return m_owner->perform_exclusive([&]
                 {
-                    ++m_current;
-                }
+                    this->check_version(*m_owner);
+                    if (m_current != m_end)
+                    {
+                        ++m_current;
+                    }
 
-                return HasCurrent();
+                    return m_current != m_end;
+                });
             }
 
             uint32_t GetMany(array_view<T> values)
             {
-                return GetMany(values, typename std::iterator_traits<iterator_type>::iterator_category());
+                return m_owner->perform_exclusive([&]
+                {
+                    this->check_version(*m_owner);
+                    return GetMany(values, typename std::iterator_traits<iterator_type>::iterator_category());
```

Comment:
Also wondering why this is locking exclusive. I'd never noticed that `GetMany()` is modifying `m_current`. That can't be right, can it, @kennykerr? `Get()` doesn't modify it, so why would `GetMany()`?

Reply:
Yes, that's correct. The GetMany on the collection doesn't modify, but the GetMany on the iterator consumes the iterator.

Reply:
If you look at the existing internal collections - I was surprised by this behavior, too - you'll see that they do the same thing. An additional thing to note is that only vector's iterator guarantees sequential consistency here (i.e. concurrent calls to `GetMany` won't "overlap"). Map's iterator type forgoes sequential consistency while still being thread safe.

Reply:
Good to know on `GetMany` - I had no idea. I suppose that if the existing collection behaviors all do this, it's a _de facto_ behavior and should be documented as such. I created a bug on the docs here: https://github.com/MicrosoftDocs/winrt-api/issues/1732

Reply:
The idea is that a WinRT iterator can model a forward-only iterator so GetMany has to move the iterator forward.

Reply:
Agreed - it makes total sense, especially when you think about how a C# IEnumerator would need to be projected as IIterator. IEnumerator has no GetMany - just Current and MoveNext.

It's just not immediately intuitive when our brains are used to IVector.GetMany being non-mutating. Hence, the UWP doc bug. :)